topdown/tokens.go

```diff
@@ -336,14 +356,45 @@ func builtinJWTVerify(a ast.Value, b ast.Value, hasher func() hash.Hash, verify
 		return nil, err
 	}
 
+	err = token.decodeHeader()
+	if err != nil {
+		return nil, err
+	}
+	header, err := parseTokenHeader(token)
+	if err != nil {
+		return nil, err
+	}
+
 	// Validate the JWT signature
-	for _, key := range keys {
-		err = verify(key,
-			getInputSHA([]byte(token.header+"."+token.payload), hasher),
-			[]byte(signature))
 
-		if err == nil {
-			return ast.Boolean(true), nil
+	// First, check if there's a matching key ID (`kid`) in both token header and key(s).
+	// If a match is found, verify using only that key. Only applicable when a JWKS was provided.
+	if header.kid != "" {
+		if key := getKeyByKid(header.kid, keys); key != nil {
+			err = verify(key.key, getInputSHA([]byte(token.header+"."+token.payload), hasher), []byte(signature))
+
+			return ast.Boolean(err == nil), nil
+		}
+	}
+
+	// If no key ID matched, try to verify using any key in the set
+	// If an alg is present in both the JWT header and the key, skip verification unless they match
+	for _, key := range keys {
+		if key.alg == "" {
+			// No algorithm provided for the key - this is likely a certificate and not a JWKS, so
+			// we'll need to verify to find out
+			err = verify(key.key, getInputSHA([]byte(token.header+"."+token.payload), hasher), []byte(signature))
+			if err == nil {
+				return ast.Boolean(true), nil
+			}
+		} else {
+			if header.alg != key.alg {
+				continue
+			}
+			err = verify(key.key, getInputSHA([]byte(token.header+"."+token.payload), hasher), []byte(signature))
+			if err == nil {
+				return ast.Boolean(true), nil
+			}
 		}
 	}
```

Comment:
Should we consider also skipping any key here that has a `kid` that doesn't match the `kid` of the JWT?
This would step outside the realm of just being an optimization; but it might be the correct thing to do ..

Reply:
We would still try keys without `kid`, though. 
They'd have broken JWT verification if they expect to use a key that has a different `kid` than the JWT. But yes, this would be a change in behavior; which may be frowned upon.

Reply:
Ah, I see what you mean. Yeah we probably _should_ do that at some point. Might be a good idea to have an `opa1.0`-labeled ticket for everything around JWT, JWKS and certificates (also in the OPA configuration) which would be good fixes but probably not worth doing in a minor release.